### PR TITLE
Support swap fee

### PR DESCRIPTION
**Pull request by hungdoansy**
# What 
KyberSwap will implement Swap Fee on these low-trading-volume chains: Oasis, Velas, Aurora, Cronos
| Swap Form | Confirm Modal |
| --- | --- |
| image | image |

# How
- Tokens are divided into 2 tiers. Tier 1 includes stable tokens and native token. Others fall into Tier 2.
Swap fee and which token fee is taken from will be calculated by tiers of both tokens, where it's a stable coin swap, and scores of them. Score of a token is calculated by the quantity of native token which can be obtained by swapping $1000 worth of that token.
- Upon a GET route request, if the response has extraFee (indicating that fee is applied in this swap), calculate how much fee (in token and in usd), and put this intel (as `routeSummary.fee`) inside detailed route summary for later use
- Fee is once again calculated in Confirm Swap modal. There's badge to indicate any changes made to fee.

# To do
- [x] Update BFF_API on production

### CODE REVIEW

File: src/components/SwapForm/OutputCurrencyPanel.tsx [hunk 1]
```diff
@@ -76,12 +79,17 @@ const OutputCurrencyPanel: React.FC<Props> = ({
             text={
               <Text fontSize={12}>
                 <Trans>
-                  This is the estimated output amount. Do review the actual output amount in the confirmation stage.
+                  This is the estimated output amount. It is inclusive of any applicable swap fees. Do review the actual
```
Comment (nhd98z) [inline src/components/SwapForm/OutputCurrencyPanel.tsx:82]: Tooltip should check CHAINS_SUPPORT_FEE_CONFIGS too.

**Review by nhd98z — CHANGES_REQUESTED** [review]

File: src/components/SwapForm/hooks/useGetRoute.ts [hunk 2]
```diff
@@ -18,6 +19,8 @@ export type ArgsGetRoute = {
   parsedAmount: CurrencyAmount<Currency> | undefined
   currencyIn: Currency | undefined
   currencyOut: Currency | undefined
+
+  // TODO: remove this
```
Comment (nhd98z) [inline src/components/SwapForm/hooks/useGetRoute.ts:23]: TODO?

File: src/components/SwapForm/hooks/useGetSwapFeeConfig.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  const { isEnableAuthenAggregator } = useKyberswapGlobalConfig()
+
+  const [triggerGetTokenScoreQuery] = tokenApi.useLazyGetTokenScoreQuery()
+
+  const getSingleTokenScore = useCallback(
+    async (chainId: ChainId, tokenAddress: string): Promise<string | undefined> => {
+      const now = Math.floor(Date.now() / 1000)
+
+      const { score, savedAt } = getTokenScoreFromLocal(chainId, tokenAddress) || {}
+      if (score && savedAt && now - savedAt < 86400) {
```
Comment (nhd98z) [inline src/components/SwapForm/hooks/useGetSwapFeeConfig.ts:66]:
```suggestion
      if (score && savedAt && now - savedAt < TOKEN_SCORE_TTL) {
```

File: src/utils/fee.ts [hunk 4]
```diff
@@ -35,23 +35,63 @@ export function getFormattedFeeAmountUsd(trade: Aggregator, feeConfig: FeeConfig
   return '--'
 }
 
-export const getFormattedFeeAmountUsdV2 = (rawAmountInUSD: number, feeAmount?: string) => {
-  if (feeAmount) {
-    const amountInUsd = new Fraction(
-      parseUnits(toFixed(rawAmountInUSD), RESERVE_USD_DECIMALS).toString(),
+export const calculateFee = (
```
Comment (nhd98z) [inline src/utils/fee.ts:38]: No need to calculate this in FE!

File: src/components/SwapForm/OutputCurrencyPanel.tsx (hunk 1 above)
Comment (hungdoansy) [reply src/components/SwapForm/OutputCurrencyPanel.tsx]: Done

File: src/components/SwapForm/SwapActionButton/SwapOnlyButton.tsx [hunk 5]
```diff
@@ -110,6 +133,14 @@ const SwapOnlyButton: React.FC<Props> = ({
       gasUsd: routeSummary?.gasUsd,
       inputAmount: routeSummary?.parsedAmountIn,
       priceImpact: routeSummary?.priceImpact,
+      feeInfo: routeSummary?.extraFee?.chargeFeeBy
+        ? {
+            chargeTokenIn: routeSummary.extraFee.chargeFeeBy === 'currency_in',
+            tokenSymbol: currencyToTakeFee?.symbol || '',
+            feeUsd: routeSummary.extraFee.feeAmountUsd || feeAmountUsd,
```
[the remark this answers is not in the document]
Comment (hungdoansy) [reply src/components/SwapForm/SwapActionButton/SwapOnlyButton.tsx]: Done

File: src/components/SwapForm/SwapActionButton/SwapOnlyButton.tsx [hunk 6]
```diff
@@ -164,14 +195,22 @@ const SwapOnlyButton: React.FC<Props> = ({
           priceImpact: routeSummary?.priceImpact,
           outputAmountDescription,
           currentPrice,
+          feeInfo: routeSummary?.extraFee?.chargeFeeBy
+            ? {
+                chargeTokenIn: routeSummary.extraFee.chargeFeeBy === 'currency_in',
+                tokenSymbol: currencyToTakeFee?.symbol || '',
+                feeUsd: routeSummary.extraFee.feeAmountUsd || feeAmountUsd,
```
[the remark this answers is not in the document]
Comment (hungdoansy) [reply src/components/SwapForm/SwapActionButton/SwapOnlyButton.tsx]: Done

File: src/components/SwapForm/hooks/useGetRoute.ts (hunk 2 above)
Comment (hungdoansy) [reply src/components/SwapForm/hooks/useGetRoute.ts:23]: Will do in the next PR. To remove referral program completely

File: src/utils/fee.ts (hunk 4 above)
Comment (hungdoansy) [reply src/utils/fee.ts]: Done

File: src/components/SwapForm/hooks/useGetSwapFeeConfig.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+} from 'constants/tokens'
+import { useKyberswapGlobalConfig } from 'hooks/useKyberSwapConfig'
+import { getTokenScore as getTokenScoreFromLocal, saveTokenScore } from 'utils/tokenScore'
+
+export type SwapFeeConfig = {
+  token: string
+  feeBips: number
+}
+
+const timeoutReject = (ms: number, msg = 'timeout') => {
```
[the remark this answers is not in the document]
Comment (hungdoansy) [reply src/components/SwapForm/hooks/useGetSwapFeeConfig.ts]: Done

File: src/components/SwapForm/hooks/useGetSwapFeeConfig.ts (hunk 3 above)
Comment (hungdoansy) [reply src/components/SwapForm/hooks/useGetSwapFeeConfig.ts]: Done

**Review by nhd98z — DISMISSED** [review]

File: .env.production [hunk 8]
```diff
@@ -37,3 +37,5 @@ VITE_KYBER_DAO_STATS_API=https://kyberswap-dao-stats.kyberengineering.io
 VITE_ENV=production
 VITE_OAUTH_CLIENT_ID=8a95ced8-1530-45d0-bc15-a5614d1d8d08
 VITE_PRICE_ALERT_API=https://price-alert.kyberswap.com/api
+
+VITE_BFF_API=
```
Comment (viet-nv) [inline .env.production:41]: has BE release yet?

Comment (hungdoansy) [reply .env.production:41]: Not yet. I added a task

File: src/components/SwapForm/SwapActionButton/SwapOnlyButton.tsx [hunk 9]
```diff
@@ -107,6 +107,14 @@ const SwapOnlyButton: React.FC<Props> = ({
       gasUsd: routeSummary?.gasUsd,
       inputAmount: routeSummary?.parsedAmountIn,
       priceImpact: routeSummary?.priceImpact,
+      feeInfo: routeSummary?.fee
+        ? {
+            chargeTokenIn: routeSummary.extraFee.chargeFeeBy === 'currency_in',
```
Comment (nguyenhoaidanh) [inline src/components/SwapForm/SwapActionButton/SwapOnlyButton.tsx:112]: duplicate vs line 112 => write a utils function?

File: src/hooks/useSwapCallbackV3.ts [hunk 10]
```diff
@@ -84,22 +84,31 @@ const useSwapCallbackV3 = (isPermitSwap?: boolean) => {
           slippageSetting: allowedSlippage ? allowedSlippage / 100 : 0,
           priceImpact: priceImpact && priceImpact > 0.01 ? priceImpact.toFixed(2) : '<0.01',
           isPermitSwap,
+          feeInfo: routeSummary?.fee
+            ? {
+                chargeTokenIn: routeSummary.extraFee.chargeFeeBy === 'currency_in',
```
Comment (nguyenhoaidanh) [inline src/hooks/useSwapCallbackV3.ts:89]: duplicate

File: src/components/SwapForm/SwapActionButton/SwapOnlyButton.tsx (hunk 9 above)
Comment (hungdoansy) [reply src/components/SwapForm/SwapActionButton/SwapOnlyButton.tsx]: Nice. I fixed it

File: src/hooks/useSwapCallbackV3.ts (hunk 10 above)
Comment (hungdoansy) [reply src/hooks/useSwapCallbackV3.ts:89]: Nice. I fixed it

File: .env.production (hunk 8 above)
Comment (hungdoansy) [reply .env.production]: Updated API for prod

**Review by viet-nv — APPROVED** [review]

**Review by nhd98z — APPROVED** [review]

**Pull request merged**